### PR TITLE
bpo-42111: Make the xxlimited module an example of best extension module practices

### PR DESCRIPTION
**Pull request by encukou**
- Copy existing xxlimited to xxlimited53 (named for the limited API version it uses)
- Build both modules, both in debug and release
- Test both modules

https://bugs.python.org/issue42111

Automerge-Triggered-By: GH:encukou

### CODE REVIEW

File: setup.py
Comment (vstinner) [inline setup.py]: Hum, it should be 3.5 rather than 3.10 no?

Why Python 3.5 in specific and not 3.6 for example? Nowadays, 3.6 became the minimum version of many projects, and it's also the Python 3 version of RHEL8 ;-)

File: setup.py
Comment (vstinner) [inline setup.py:1813]: Why not also building with the limited C API?

File: setup.py
Comment (vstinner) [inline setup.py:1808]: Is it supposed to be updated when the python version is increased? If yes, maybe PEP 101 should be updated for release managers. (after this PR is merged)

File: Modules/xxlimited_35.c
Comment (vstinner) [inline Modules/xxlimited_35.c:239]: Is a type spec used still used after a PyType_FromSpec() call? If not, maybe move type specs definition inside the function, to avoid this workaround?

File: Modules/xxlimited_35.c
Comment (vstinner) [inline Modules/xxlimited_35.c:252]: Would you mind to check for error? If < 0: Py_DECREF(ErrorObject); goto fail;

File: Modules/xxlimited_35.c
Comment (vstinner) [inline Modules/xxlimited_35.c:160]: Maybe mention that "O#" format is invalid and raises SystemError.

File: Lib/test/test_xxlimited.py
Comment (vstinner) [inline Lib/test/test_xxlimited.py:79]: Maybe add: `if __name__ == "__main__": unittest.main()` (not sure of the code, look at the other files). Some people like to run directly a test file, without "python -m test".

File: Modules/xxlimited.c
Comment (vstinner) [inline Modules/xxlimited.c:76]: It works if you pass a state, and the type from the state, no?

File: Modules/xxlimited.c
Comment (vstinner) [inline Modules/xxlimited.c]: Please add error checking: if < 0, return -1.

File: Modules/xxlimited.c
Comment (vstinner) [inline Modules/xxlimited.c:324]: maybe the spec could be defined here to use the tp_base member?

File: Modules/xxlimited.c
Comment (vstinner) [inline Modules/xxlimited.c]: Hum, it should be DECREF no? Try: ./python -m test -R 3:3 test_xxlimited

File: Modules/xxlimited.c
Comment (vstinner) [inline Modules/xxlimited.c]: PEP 7 suggests:

```suggestion
xx_traverse(PyObject *module, visitproc visit, void *arg)
{
```

if I recall properly.

File: Modules/xxlimited_35.c
Comment (vstinner) [inline Modules/xxlimited_35.c:3]: If you prefer Python 3.5 over Python 3.6, maybe add a note that C99 is not used, since C99 is only supported since Python 3.6. I suggest that since I wanted to use you to defined types with C99 ".member = ..." syntax :-)

File: Lib/test/test_xxlimited.py
Comment (vstinner) [inline Lib/test/test_xxlimited.py:4]: nitpick: PEP 8 suggests two empy lines after imports, if I recall properly :-D

**Review by vstinner — COMMENTED** [review]
I'm curious why you picked Python 3.5 rather than 3.6 :-)

**Comment by encukou** [comment]
`xxlimited_35.c` is a copy of the historical code; I don't want to change it.
But I'll definitely look at your other suggestions!

File: setup.py
Comment (encukou) [reply setup.py]: Good catch!

The module uses limited API from Python 3.5. The corresponding stable ABI needs to be still compatible with current Python versions.

File: setup.py
Comment (encukou) [reply setup.py:1813]:
```
./Include/object.h:61:2: error: #error Py_LIMITED_API is incompatible with Py_DEBUG, Py_TRACE_REFS, and Py_REF_DEBUG
```

I guess that can be fixed, but that's out of scope here (and I don't know what the situation around this is on Windows)

File: setup.py
Comment (encukou) [reply setup.py:1808]: No. When new things are needed, we should copy `xxlimited` to e.g. `xxlimited_310`, add tests, and update `xxlimited` to the newest version. If new stuff isn't needed in a release, this can stay on the old limited API version.

File: Lib/test/test_xxlimited.py
Comment (encukou) [reply Lib/test/test_xxlimited.py:79]: I think that is unnecessary; `./python -m test` should work just fine.
(Or is the requirement for the `if __name__ == "__main__":` hack in the devguide?)

File: Modules/xxlimited.c
Comment (encukou) [reply Modules/xxlimited.c:76]: I'd like to leave this out of scope for this PR.

File: Lib/test/test_xxlimited.py
Comment (encukou) [reply Lib/test/test_xxlimited.py:4]: The following lines also imports ;)

File: Modules/xxlimited.c
Comment (encukou) [reply Modules/xxlimited.c]: Thanks for the catch!

File: Modules/xxlimited.c
Comment (encukou) [reply Modules/xxlimited.c:324]: That doesn't quite work on Windows IIRC. But PyType_FromModuleAndSpec can take a single type nowadays, I'll switch to that.

Comment (vstinner) [reply Modules/xxlimited.c:324]: > That doesn't quite work on Windows IIRC

If you define the spec variable in the exec function, it should work with any compiler. The issue is only for "global" variables.

File: Lib/test/test_xxlimited.py
Comment (vstinner) [reply Lib/test/test_xxlimited.py:79]: No, it's not a requirement. It's just that other test_xxx.py files have it. You can leave this file as it is.

File: setup.py
Comment (vstinner) [reply setup.py:1813]: Oh. Maybe it was needed when the debug build was not ABI compatible with the release build. It has been fixed in Python 3.8. But yeah, it can be addressed in a separated PR.

File: setup.py
Comment (vstinner) [reply setup.py:1808]: Ok.

**Review by vstinner — APPROVED** [review]
LGTM.

I checked `./python -m test test_xxlimited -R 3:3`: there is no leak.

**Comment by encukou** [comment]
Thank you for the review!

**Pull request merged**
